feat: 내 책장의 책 목록 DB에서 가져오기 및 선택한 책 삭제

Ah-ae opened the pull request:
## 주요 변경 사항 
- 사용자별 책장의 책 전체 목록 DB에서 조회하여 /mine 페이지에서 책 표지 섬네일 렌더링 (prisma studio를 사용해 DB에 먼저 데이터 직접 넣어둔 후 해당 데이터에 접근)
- 상단 Header 컴포넌트 마크업 작성 
- Header 컴포넌트의 우측 `편집` 또는 `완료` 버튼으로 `view`, `edit` 모드 전환
- edit 모드일 때 사용자가 책 표지 섬네일을 클릭할 때 조건부 렌더링 구현 (dimmed overlay 및 빨간색 체크 아이콘) 

### 추후 작업 예정 
- deleteBooks 서버 액션 함수 작성 
- Header 컴포넌트와 책 목록 사이에 Search 컴포넌트를 만들어 query에 따라 책 전체 목록 중 필터링한 목록 보여주기 
- view 모드일 때 표지 섬네일 클릭시 책 상세 페이지로 이동 (현재 404) 
- edit 모드일 때 드래그앤드롭으로 책 목록 순서 재정렬 구현 
- edit 모드일 때 재정렬 버튼 클릭시 '등록순', '등록순(역순)', '제목순', '저자순' 등 재정렬 기능 구현 

### 이슈 
- 전역 상태를 사용하지 않고 사용자가 선택한 아이템(`selectedItems`) 데이터를 여러 컴포넌트에서 공유할 수 있도록 하는 게 어려웠음 
  * 처음에는 session storage에 selectedItems: number[] 값을 저장하려고 했는데 해당 값을 onClick 이벤트 핸들러에서만 실시간으로 접근 가능해서 휴지통 아이콘 색상 조건부 렌더링이나 삭제 버튼의 disabled 처리를 구현할 수 없었음 - [94a5e5d](94a5e5d055f28d9cb627251445756336e81b7440)
  : ⚠️ **session storage 값은 왜 실시간으로 읽지 못하는데 url params는 읽을 수 있는지 digging 필요**
- https://github.com/Ah-ae/what-are-you-reading/issues/11

### 스크린샷 
Prisma Studio `Book` 테이블
<img width="1960" alt="스크린샷 2024-06-30 오후 12 38 39" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/5f7e3af4-b7fe-40ee-998a-c95463325de5">

라이트 모드
<img width="400" alt="스크린샷 2024-06-30 오후 12 32 48" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/0f85f1f7-530d-4eec-982c-35ff381065e1">
<img width="400" alt="스크린샷 2024-06-30 오후 12 33 00" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/b204b9fb-fdf0-472c-9450-f8c0bc73fdf1">
<img width="400" alt="스크린샷 2024-06-30 오후 12 33 11" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/752e5b6b-be9b-4514-b928-2338deea3d9e">

다크 모드
<img width="400" alt="스크린샷 2024-06-30 오후 12 33 59" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/2f8b4a4a-9a9b-4e57-8a29-e1a054d0b6e6">
<img width="400" alt="스크린샷 2024-06-30 오후 12 35 28" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/a27a018b-d230-4044-a8f4-349068f59e81">
<img width="400" alt="스크린샷 2024-06-30 오후 12 33 38" src="https://github.com/Ah-ae/what-are-you-reading/assets/115610668/d06435e0-3a1b-4a30-854f-a0011e884cdc">

Ah-ae committed:
feat: 사용자별 책장의 책 전체 목록 조회

Ah-ae committed:
update: 하단 네비게이션 시맨틱 마크업으로 변경 (div → nav)

Ah-ae committed:
feat: Header 컴포넌트 작성

Ah-ae committed:
feat: view 또는 edit 모드에 따라 조건부 렌더링

Ah-ae committed:
feat: 섬네일 아이템 클릭 토글에 따른 UI 조건부 렌더링 및 selectedItems 업데이트

Ah-ae committed:
rename: component 하위 디렉토리 생성

Ah-ae committed:
fix: sessionStorage.removeItem 관련 로직 useEffect로 이동

https://nextjs-faq.com/browser-api-client-component#moving-logic-to-useeffect

Ah-ae committed:
update: session storage 대신 url params에 selectedItems 값 저장

- session storage에 저장시 Header 컴포넌트에서 실시간으로 값에 접근할 수 없었음
- onClick 이벤트 핸들러에서서만 실시간 접근 가능했음

Ah-ae committed:
update: url param 바뀌어도 뒤로가기에 영향 미치지 않게 replace로 대체

Ah-ae committed:
package: clsx 설치

Ah-ae committed:
feat: clsx 사용해 휴지통 아이콘 색상 조건부 렌더링

Ah-ae committed:
fix: '완료' 클릭해 currentMode 변경시 로컬 상태 isSelected 초기화 로직 추가

Ah-ae committed:
update: DeleteBooks 추출

Ah-ae committed:
update: tailwind.config에서 main-theme 색상 HEX 값에서 colors.green[800] 으로 변경

Ah-ae commented:
추가 예정 작업들을 위한 칸반 보드 작성:
https://github.com/users/Ah-ae/projects/3

Ah-ae committed:
refactor: getUser 함수 data.ts 로 이동

Ah-ae committed:
update: BookThumbnail 컴포넌트 주석 추가 및 일부 변수명 변경

Ah-ae committed:
update: import 경로 절대경로로 수정

Ah-ae committed:
rename: components → ui 로 디렉토리명 변경

Ah-ae committed:
rename: utils/login → utils/auth 파일명 변경

Ah-ae committed:
refactor: useBookThumbnail hook 추출

Ah-ae committed:
fixup! refactor: getUser 함수 data.ts 로 이동

yun-coco commented on app/(auth)/signup/actions.ts:
```diff
@@ -5,7 +5,7 @@ import bcrypt from 'bcrypt';
+import { login } from '@/utils/auth';
 
 const checkUserName = (username: string) => {
   for (const disallowedWord of DISALLOWED_USERNAME_LIST) {
```
Array.some() 메소드 사용하면 어떨까요?
  
```
   const checkUserName = (username: string): boolean => {
      return !DISALLOWED_USERNAME_LIST.some(disallowedWord => 
        username.toLowerCase().includes(disallowedWord)
      );
    }; 
```

Ah-ae replied:
오호,, Array.some(), every() 메소드를 한 번도 사용해 본 적이 없어서 생각을 못해봤어요...!

Ah-ae committed:
refactor: checkUsername 함수 Array.some() 메소드로 변경

- PR 피드백 반영 https://github.com/Ah-ae/what-are-you-reading/pull/12#discussion_r1667322810
- 함수형 프로그래밍의 장점을 살릴 수 있고, 코드의 간결함과 가독성이 높아짐
